File: knapsack_benchmark/knapsack.py

import sys
if sys.hexversion < 0x3000000: raise RuntimeError("expecting python v3+ instead of %x" % sys.hexversion)

import array
import numpy as np
import time

# ----------------------------------------------------------------------------
'''
    this is an almost one-for-one translation of the Julia version, see knapsack.jl
    for more comments.
    
    running this benchmark from command line:
    
    >python3 -O ...knapsack.py
'''

class item(object):
    def __init__(self, value, weight):
        self.value = value
        self.weight = weight
        
    def __repr__(self):
        return "(" + str(self.value) + ", " + str(self.weight) + ")"
# ............................................................................
        
def opt_value(W, items):    
    n = len(items)
    
    V = [0 for i in range(W)]
    V_prev = [0 for i in range(W)]

    # use the next 2 lines to use python "native" arrays instead:
    # (for me, this choice is slower 2x than plain python lists)
    
#     V = array.array('q', [0 for i in range(W)])
#     V_prev = array.array('q', [0 for i in range(W)])

    # use the next 2 lines to use numpy arrays instead:
    # (for me, this choice is nearly 2x slower)

#     V = np.array([0 for i in range(W)], dtype="i8")
#     V_prev = np.array([0 for i in range(W)], dtype="i8")
    
    for w in range(items[0].weight, W + 1):
        V[w - 1] = items[0].value;
        
    for j in range(1, n):
        V, V_prev = V_prev, V
        item_j = items[j]
        for w in range(1, W + 1):
            V_without_item_j = V_prev[w - 1]
            V_allow_item_j = (V_without_item_j if w < item_j.weight
                else (item_j.value + (V_prev[w - 1 - item_j.weight] if w != item_j.weight
                    else 0)))
            
            V[w - 1] = max(V_allow_item_j, V_without_item_j)
            
    return V[W - 1]             
    
# ............................................................................

# some contortions are needed in python to ensure uint64_t arithmetic:

_13 = np.uint64(13)
_7  = np.uint64(7)
_17 = np.uint64(17)

def xorshift_rand(seed):
    assert seed != 0
    x = np.uint64(seed)
    
    def _next():
        nonlocal x
        x ^= (x << _13);
        x ^= (x >> _7);
        x ^= (x << _17);

        return int(x);
    
    return _next

def make_random_data(W, seed):
    assert W > 1000
    n = W // 100
    rng = xorshift_rand(seed)
    
    items = []
    for i in range(n):
        v = rng() % 1000
        w = 1 + rng() % (2 * W)
         
        items.append(item(v, w))
    
    return W, items

# ............................................................................

def run(repeats = 5):
        
    times = [0.0 for i in range(repeats)]
    seed = 12345
    
    for W in [5000, 10000, 20000, 40000, 80000]:
        for repeat in range(repeats):
            seed += 1
            W, items = make_random_data(W, seed)
            start = time.time_ns ()
            
            V = opt_value(W, items)
            
            stop = time.time_ns()
            times[repeat] = (stop - start) / 1e9
#             print("V = %d, time = %f" % (V, times[repeat]))
            
        times.sort()
        print("python, %d, %f" % (W, times[repeats // 2]))

# ............................................................................

if __name__ == '__main__':
    run ()

# ----------------------------------------------------------------------------

File: knapsack_benchmark/Knapsack.java

import java.util.Arrays;

//----------------------------------------------------------------------------
/*
	this is an almost one-for-one translation of the Julia version, see knapsack.jl
	for more comments.

	running this benchmark from command line:

	>mkdir bin; javac -d bin Knapsack.java
	>java -cp bin Knapsack
*/
public class Knapsack
{	
	static final class Item
	{
		Item(final long value, final long weight)
		{
			this.value = value;
			this.weight = weight;
		}
		
		final long value;
		final long weight;
		
	} // end of nested class
	//........................................................................
	/*
	 * some types are changed to 'int' in this version because in Java
	 * you can't index into arrays with types wider than 'int'
	 */
	static long optValue(final int W, final Item[] items)
	{
		final int n = items.length;
		
		long[] V = new long [W];
		long[] V_prev = new long [W];
		
	    for (int w = (int)items[0].weight; w <= W; ++ w)
	        V[w - 1] = items[0].value;
	    
	    for (int j = 1; j < n; ++ j)
	    {
	        long[] temp = V; V = V_prev; V_prev = temp;

	        final Item item = items[j];

	        for (int w = 1; w <= W; ++ w)
	        {
	            final long V_without_item_j = V_prev[w - 1];
	            final long V_allow_item_j = (w < item.weight
	                ? V_without_item_j
	                : (item.value + (w != item.weight ? V_prev[w - 1 - (int)item.weight] : 0)));

	            V[w - 1] = Math.max(V_allow_item_j, V_without_item_j);
	         }
	    }
	    
	    return V[W - 1];
	}
	//........................................................................
	
	static final class XorshiftRandom
	{
		XorshiftRandom(final long seed)
		{
			x = seed;
		}
		
		public long next()
		{
			x ^= (x << 13);
			x ^= (x >>> 7);
			x ^= (x << 17);
			
			return x;
		}
		
		private long x;
	}
	//........................................................................
	
	static final class Problem
	{
		Problem (final int W, final Item[] items)
		{
			this.W = W;
			this.items = items;
		}
		
		final int W;
		final Item[] items;
		
	} // end of nested class
	
	static Problem makeRandomData(final int W, final long seed)
	{
		assert W > 1000;
		final int n = W / 100;
		final XorshiftRandom rng = new XorshiftRandom(seed);
		
		final Item[] items = new Item[n];
		
		for (int i = 0; i < n; ++ i)
		{
			final long v = Long.remainderUnsigned(rng.next(), 1000);
			final long w = 1 + Long.remainderUnsigned(rng.next(), 2 * W);
			
//			System.out.println("v = " + v + ", w = " + w);
			items[i] = new Item(v, w);
		}
		
		return new Problem(W, items);
	}
	//........................................................................
	
	static void run(final int repeats)
	{
		final double[] times = new double[repeats];
		long seed = 12345;
		
		final int[] Ws = { 5_000, 10_000, 20_000, 40_000, 80_000 };
		for (int W : Ws)
		{
			for (int repeat = 0; repeat < repeats; ++ repeat)
			{
				final Problem spec = makeRandomData(W, seed += 1);
				final long start = System.nanoTime();
				
				final long V = optValue(spec.W, spec.items);
		
				final long stop = System.nanoTime();
				times[repeat] = (stop - start) / 1e9;
//				System.out.println("W = " + W + ", V = " + V + ", time: " + times[repeat]);
			}
			
			Arrays.sort(times);
			System.out.println("java, " + W + ", " + times[repeats / 2]);
		}
	}
	//........................................................................
	
	public static void main(String[] args)
	{
		run(5);
	}
}
//----------------------------------------------------------------------------
